Using C libs: Update doc on parsing/receiving data

hamishwillee opened the pull request:
Primarily, this adds more detailed instructions on decoding packets, and from them messages.

hamishwillee commented on en/mavgen_c/README.md:
````diff
@@ -59,16 +60,164 @@ When compiling the project, we recommend that you specify the top-level output d
 ```
 
-## Multiple Streams ("channels") {#streams}
+## Multiple Streams ("channels") {#channels}
````
FYI, this section could do with a little more. I don't find this explanation usefully explains when I would use a new channel, and when not. Are there examples of what make sense as a channel?

julianoes commented:
I'm putting this on my todo list to review some time next week.

hamishwillee commented:
@peterbarker, FYI too - appreciate it if you have any feedback on these addition to docs on parsing/receiving packets in the C library.

peterbarker commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,164 @@ When compiling the project, we recommend that you specify the top-level output d
+The C MAVLink library utilizes a "channel" metaphor to allow for simultaneous processing of multiple, independent MAVLink streams in the same program.
+All receiving and transmitting functions provided by this library require a channel, and it is important to use the correct channel for each operation.
 
+Up to 4 channels can be defined on MCU/ARM7 while up to 16 may be defined on Linux/Windows.
```
Perhaps a "by default" and `s/can be/are/`.  ArduPilot specifies more channels.

peterbarker replied:
... also, `MCU/ARM7` is garbage, it's "everything other than xyzzy":

```
#ifndef MAVLINK_COMM_NUM_BUFFERS
#if (defined linux) | (defined __linux) | (defined  __MACH__) | (defined _WIN32)
# define MAVLINK_COMM_NUM_BUFFERS 16
#else
# define MAVLINK_COMM_NUM_BUFFERS 4
#endif
```

Poor MacOS users, always forgotten :-)

hamishwillee replied:
Thanks, will fix. 

> Poor MacOS users, always forgotten :-)

Yes, because they suck :-)

peterbarker commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,164 @@ When compiling the project, we recommend that you specify the top-level output d
-MAVLink reception is then done using `mavlink_helpers.h:mavlink_parse_char()`.
+MAVLink reception/decoding is done in a number of phases: 
+1. Parse the incoming stream into objects representing each packet (`mavlink_message_t`).
+1. Decode the MAVLink message contained in the packet payload into a C struct (that has fields mapping the original XML definition).
```
Sequence error?

hamishwillee replied:
No, markdown uses `1.` to indicate any numbered bullet, and provides autonumbering.

peterbarker commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,164 @@ When compiling the project, we recommend that you specify the top-level output d
+
+### Parsing Packets
+
+The `mavlink_parse_char()` convenience function ([mavlink_helpers.h](https://github.com/mavlink/c_library_v2/blob/master/mavlink_helpers.h)) is used to parse incoming MAVLink data.
```
elipsis to indicate this isn't a void-taking function...

hamishwillee replied:
Hmmm. Won't most people just assume that `mavlink_parse_char()` is a function name? Generally I add the elipsis if I am including the return type - in that case I am showing enough that people might thing that it is a signature.

peterbarker replied:
Not when you put the parentheses in.  Then it starts to look like a signature.  Could could leave the prens off.

hamishwillee replied:
I have left this as-is for now:
- The parenthesis make it immediately obvious that you're talking about a function.
- I personally never assume I'm looking at a signature unless I see a return type or it says that it is a signature.
- This pattern is used throughout the doc set.

So this should be changed globally if there is general agreement it is a problem. But for now, I don't see it as a big one.

peterbarker commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,164 @@ When compiling the project, we recommend that you specify the top-level output d
+
+> **Tip** Individual message definitions are pulled in when you call `#include <mavlink.h>`, so you don't need to include these separately.
+
+The main decoding function has the signature format `mavlink_msg_<message_name>_decode()`, and decodes the whole message into a C struct (with fields mapping to the original XML message definition).
```
signature incorrect.  elipsis are no good in this case since you've used a keyword :-)

hamishwillee replied:
Don't really understand this, but as above, it is just a name, not a signature. I've removed signature from the text now so I think the confusion is removed?

peterbarker replied:
You've *said* it has this signature - when it clearly doesn't.

hamishwillee replied:
Yes it did thanks. But following your comment I removed that - i.e. you're seeing the outdated version that I since fixed :-). 

I've further changed this to "The most useful decoding function is named with the pattern".

hamishwillee commented:
Thanks @peterbarker . Have done updates that I hope address your issues!

@julianoes Reminder!

julianoes commented on en/mavgen_c/README.md:
````diff
@@ -59,16 +60,166 @@ When compiling the project, we recommend that you specify the top-level output d
+
+The code fragment below shows the typical use of this function when reading data from a serial port (`serial`):
+```c
+#include <mavlink.h>
````
I think this include is wrong, shouldn't it be `#include <common/mavlink.h>`, or `#include <standard/mavlink.h>`?

hamishwillee replied:
Yes. Thanks!

julianoes commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,166 @@ When compiling the project, we recommend that you specify the top-level output d
+  uint8_t byte = serial.getNextByte();
+  if (mavlink_parse_char(chan, byte, &msg, &status))
+    {
+    printf("Received message with ID %d, sequence: %d from component %d of system %d", msg.msgid, msg.seq, msg.compid, msg.sysid);
```
Add `\n` at end for newline in printf. And usually we name system ID before component ID.

hamishwillee replied:
Added `\n`. Did not change ordering as this is a correct sentence, and reordering is not really part of the learning experience.

julianoes commented on en/mavgen_c/README.md:
````diff
@@ -59,16 +60,166 @@ When compiling the project, we recommend that you specify the top-level output d
+
+
+The code fragment below shows the typical use of this function when reading data from a serial port (`serial`):
+```c
````
This is C++, not  C in fact.

julianoes commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,166 @@ When compiling the project, we recommend that you specify the top-level output d
+      case MAVLINK_MSG_ID_GPS_STATUS:
+        {
+          // Get just one field from payload
+          visible_sats = mavlink_msg_gps_status_get_satellites_visible(&msg);
```
Wow, I didn't know that.

hamishwillee replied:
:-) Me either until I wrote this.

julianoes commented on en/mavgen_c/README.md:
```diff
@@ -59,16 +60,166 @@ When compiling the project, we recommend that you specify the top-level output d
+The decoder/encoder functions and ids for each message in a dialect can be found in separate header files under the dialect folder.
+The headers are named with a format including the message name (**mavlink\_msg\_*message_name*.h**)
+
+> **Tip** Individual message definitions are pulled in when you call `#include <mavlink.h>`, so you don't need to include these separately.
```
See above.

julianoes commented on en/mavgen_c/README.md:
```diff
@@ -133,3 +283,14 @@ if (mavlink_parse_char(MAVLINK_COMM_0, buf[i], &msg, &status)) {
+
+In addition, the C library is used in numerous open source systems:
+- [PX4](https://github.com/PX4/Firmware/blob/master/src/modules/mavlink/mavlink_receiver.cpp)
+- [ArduPilot](https://github.com/ArduPilot/ardupilot/blob/master/libraries/GCS_MAVLink/GCS_Common.cpp)
```
You could list the Dronecode SDK here:
https://github.com/Dronecode/DronecodeSDK/blob/develop/core/mavlink_receiver.cpp#L28-L51

hamishwillee commented:
Thanks @julianoes @peterbarker for your reviews. I have updated in line with them. I also moved the examples into the Using C section. 

I would love to better describe the channel metaphor, but this could usefully merge without that addition.

Other you OK for this to merge? If no answer will assume you are happy enough next week :-)

hamishwillee commented:
Thanks for your reviews. Merging now.
